### ISSUE TITLE
publish in v3 fails with older versions of npm, worked with v2

### ISSUE THREAD

**Issue by phun-ky** [lerna/lerna#1573]
## Expected Behavior
I expect that after upgrading to `lerna@3`, `lerna publish` works without doing changes to my npm authentication setup.

## Current Behavior
Fails with:

    lerna ERR! ENEEDAUTH You must be logged in to publish packages. Use `npm login` and try again.

It works to go into the individual packages to just `npm publish`.

## Steps to Reproduce (for bugs)
1. Upgrade to lerna 3
2. Try to push to a private repository, which worked with lerna 2

<details><summary>lerna.json</summary><p>

```json
{
  "packages": [
    "packages/*"
  ],
  "changelog": {
    "repo": "<organization>/<project>
    "cacheDir": ".changelog",
    "labels": {
      "problems: breaking change": ":boom: Breaking Change",
      "additions: feature": ":rocket: New Feature",
      "problems: bug": ":bug: Bug Fix",
      "improvements: enhancement": ":nail_care: Enhancement",
      "improvements: upgrade deps": ":arrow_up: Upgrade deps",
      "mindless: chore": ":house: Internal",
      "mindless: documentation": ":memo: Documentation"
    }
  },
  "version": "1.10.26",
  "command": {
    "init": {
      "exact": true
    },
    "publish": {
      "message": "chore(release): :bookmark: publish %s",
      "conventionalCommits": true,
      "npmClient": "npm"
    }
  }
}

```
</p></details>

Possible related to https://github.com/lerna/lerna/issues/1560

## Context
Publishing fails in our deployment cycle. 

## Your Environment

| Executable | Version |
| ---: | :--- |
| `lerna --version` | 3.0.6 |
| `npm --version`  | 6.4.0 |
| `node --version` | 8.10.0 |

| OS | Version |
| --- | --- |
| NAME | VERSION |
| Ubuntu | 16.04 LTS |

## Npm stuff

.npmrc:
```
@scope2:registry=https://waypoint.myget.org/F/scope2/npm/
//waypoint.myget.org/F/scope2/npm/:_authToken=${SCOPE2_NPM_TOKEN}

```

How I logged into the private registry:

    npm adduser --registry=https://waypoint.myget.org/F/scope2/npm/ --scope=@scope2

**Comment by phun-ky** [comment]
@daniellugn

**Comment by phun-ky** [comment]
This works if I do:

    lerna publish --no-verify-access

But there is no documentation of this. Do not close this issue unless a migration guide for 2->3 exists or a human readable documentation for this.

btw, is it possible to include this option in `lerna.json`? I tried to put it there, but with no luck:

    "command": {
      "init": {
        "exact": true
      },
      "publish": {
        "message": "chore(release): :bookmark: publish %s",
        "conventionalCommits": true,
        "npmClient": "npm",
        "noVerifyAccess": true
      }
    }

**Comment by evocateur** [comment]
>This works on MY machine if I do, but on my colleague's mac, he receives an npm auth error after all the packages have been packed: ...

@phun-ky `--no-*` prefixes are Boolean negations, so `"verifyAccess": false` is what you're looking for.

**Comment by evocateur** [comment]
After your coworker successfully runs `npm login`, does the error persist?

**Comment by evocateur** [comment]
What version of npm, etc etc, please.

**Comment by Inlustra** [comment]
@evocateur Perhaps an issue with older versions of npm? 
Our CD pipeline was on `npm v4.3.0` and `node v6.9.5`

We run a custom login script, that just logs in using regular npm login
```
Username: ***********
Password: ***********
Email: (this IS public) *****@*********.com
Logged in as *********** on ***Redacted Registry***.
```

The immediate step after that is: 

Really odd, our publish script was as follows: 
```lerna publish --yes --cd-version patch --registry ***redacted***```

But after adding ` --no-verify-registry --no-verify-access` we had a completely different issue. Maybe not be the place for this discussion mind you:

```
lerna ERR! npm publish --ignore-scripts ***********-0.0.35.tgz stderr:
npm ERR! Cannot read property 'algorithm' of undefined

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/build/.npm/_logs/2018-08-17T13_49_48_196Z-debug.log
```

We ended up having to revert back to 3.0.0-beta.20 which is working fine once again.

Looks as though this maybe be an issue with npm itself?
https://github.com/npm/npm/issues/16723

**Comment by evocateur** [comment]
@Inlustra Wonderful context, thank you! I do believe this might be related to older versions of npm.

I know there is even a still-unfixed(!) bug since v5.7.0 related to "legacy" auth and npm Enterprise (possibly all third-party?) registries. `npx npm@5.6.0 login` is necessary, even with npm 6-latest, to create an auth token for our npm Enterprise instance (once created, it is no longer necessary to run npm@5.6.0).

So, in summary, it looks like Yet Another Version Test (`npmClient === "npm" && npm -v < 5.6.0`) is needed to condense the newly-split "pack-then-publish" routine into the legacy "just-publish" method...

**Comment by phun-ky** [comment]
> ... --no-* prefixes are Boolean negations, so "verifyAccess": false is what you're looking for.

@evocateur thanks!

> After your coworker successfully runs npm login, does the error persist?

@evocateur yes, it did. We removed all traces of npm and node before we started and reinstalled node@>= 8 and npm@>=6

**Comment by evocateur** [comment]
Hrm, ok. A deeper issue, then.

**Comment by kosssi** [comment]
We have the same issue on `cozy-client` repository when Travis publish.

You can see Travis logs with [v3.1.4](https://travis-ci.org/cozy/cozy-client/builds/419240463#L681) and with [this commit](https://github.com/cozy/cozy-client/commit/76c5a09b764ded3082553b8ea712e92ce94ecee2) the deploy is ok with [v2.11.0](https://travis-ci.org/cozy/cozy-client/builds/419245291#L685).

**Comment by deporter** [comment]
I'm also having the same issue and had to use lerna v2.11.0 to fix it. 

I'm using npm v5.6.0. Is there any extra information I can give to help narrow down what is causing the issue.

**Comment by pieh** [comment]
Today we updated `lerna` from `3.0.0-rc.0` to `3.3.0` and are facing similiar problem - we are using `yarn` and default npm registry (not custom one).

We use this cmd:
```shell
lerna publish --npm-tag=next --bump=prerelease
```

lerna.json ( https://github.com/gatsbyjs/gatsby/blob/9ce045aa4f0a96802ccf5fd3a60f25afc574a8dc/lerna.json ):
```json
{
  "command": {
    "create": {
      "license": "MIT"
    },
    "publish": {
      "allowBranch": "master",
      "bump": "prerelease",
      "conventionalCommits": true,
      "message": "chore(release): Publish",
      "preid": "rc",
      "ignoreChanges": [
        "CHANGELOG.md"
      ]
    }
  },
  "loglevel": "success",
  "packages": [
    "packages/*"
  ],
  "version": "independent",
  "npmClient": "yarn",
  "useWorkspaces": true
}
```



During the publish we get:
```shell
$ lerna publish --npm-tag=next --bump=prerelease
lerna notice cli v3.3.0
Changes:
 - gatsby: 2.0.0-rc.20 => 2.0.0-rc.21

? Are you sure you want to publish these packages? Yes
lerna ERR! ENEEDAUTH You must be logged in to publish packages. Use `npm login` and try again.
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
➜  gatsby git:(master) npm whoami
pieh
```
we are actually logged in and without any additional `npm login` can actually publish packages manually with `yarn publish`.

But biggest problem is that publish commit (and tags) get pushed to repository (without actually publishing packages) - https://github.com/gatsbyjs/gatsby/commit/938d168acb72b857410b98a0ec4d2a9b65adbd37 and we have to follow that with manual `yarn publish --tag next` to actually publish package. Any clues?

**Comment by evocateur** [comment]
* I don't recommend using yarn for publishing. There are far too many edge cases. I'm likely going to switch lerna to libnpmpublish soon.
* bump is a positional argument, not a flag.
* If the commits are made successfully, you can run `lerna publish from-git` to rerun only the publish steps (skipping git commit and tagging).

**Comment by pieh** [comment]
> I don't recommend using yarn for publishing. There are far too many edge cases. I'm likely going to switch lerna to libnpmpublish soon.

I did a little digging and my error seems to come from here - https://github.com/lerna/lerna/blob/5da13190852897ac37349a28a0b24470ec7bd833/commands/publish/lib/verify-npm-package-access.js#L55-L60
which seems to be checking response from `npm access ls-packages --fetch-retries=0 --loplevel=http`
which works fine when done in terminal/console - so this is confusing why it only happens in lerna 

> bump is a positional argument, not a flag.

Right, we use `bump` in cmd too, so even if it's no-op it shouldn't be a problem here? Actual version bumping works as expected (just wanted to include lerna.json to provide as much context as possible)

> If the commits are made successfully, you can run `lerna publish from-git` to rerun only the publish steps (skipping git commit and tagging).

I'll keep that in mind, but should `lerna` actually push "publish" commit when no package were really published?

**Comment by evocateur** [comment]
> should `lerna` actually push "publish" commit when no package were really published?

I've personally gone both ways on this question, depending on the context. However, it's always been the case that `npm publish` _does not care_ what the state of git is, whether it's been pushed or not, etc. `npm version` of course will blow up if your tree is dirty, but that's a separate concern.

Nowadays I err on the side of pushing early, then publishing, largely because it's _more_ exasperating to find a bug in a new version of a package, only to find that the "latest" source code does not match the packaged code (basically, they forgot to `git push`). Publishing packages to a registry is also (by _far_) the more risky operation, whereas the git operations are quite fast and fairly reliable.

You can always pass `--no-push` and `git push` manually, for example:
```
npx lerna publish --no-push && git push
```

**Comment by pieh** [comment]
Thanks for your time @evocateur responding to my comments. 3rd attempt actually did work without any changes (previous 2 attempts were done on 2 different machines by 2 different people) - so it seems this was unlucky unresponsive (or badly responsing?) npm registry 2 times in a row (even if manual publishing worked in between those 2 tries). I'll use `lerna publish from-git` instead of manual publishes next time. Will get bigger sample size to see if this problems will happen often and will consider using `--no-verify-access` which would be shame because we updated lerna largely because of access verification.

**Comment by evocateur** [comment]
@pieh I appreciate the persistence! I definitely would like to figure out how to make lerna more robust for large projects like Gatsby. :)

**Comment by nolazybits** [comment]
I have the same problem
lerna.json
```json
{
  "packages": [
    "packages/*"
  ],
  "version": "0.4.5",
  "npmClient": "yarn",
  "useWorkspaces": true,
  "command": {
    "init": {
      "exact": true
    },
    "publish": {
      "allowBranch": [
        "master"
      ],
      "conventional-commits": true,
      "message": "chore(release): publish %v",
      "npmClient": "yarn"
    }
  }
}
```

I am running the command from a container extending node-alpine:10.7.0 and get 
```
lerna info auto-confirmed
lerna info publish Publishing packages to npm...
lerna ERR! ENEEDAUTH You must be logged in to publish packages. Use `npm login` and try again.
error Command failed with exit code 1.
```

going in the container and running `npm whoami` yields my username. `npm token list` will ist my tokens so I am sure I am logged in.

I have bypass the error adding the `--no-verify-access` flag.

Also, it SEEMS (I played so much with this issue, but I'm almost certain that was the case) that this error might come when I run `yarn publish-all` which in turn will run `lerna publish --yes --exact --conventional-commits`. If I run the `lerna publish --yes --exact --conventional-commits` straights it works.

**Comment by mpeyper** [comment]
```
Node:  10.12.0
NPM:   6.4.1
Lerna: 3.4.3
Yarn:  No
```
I ran into the "Cannot read property 'algorithm' of undefined" error and, like @darkobits, discovered that `lerna` was trying to use `npm: 5.1.0` when publishing.  `npm ls npm` showed that the older npm version was coming from another dependency (`gitbook-cli` in this case).

I ran `npm i npm@^6` and tried publishing again and this time it there was no issues.  Hope this helps someone.

**Comment by darkobits** [comment]
🤦‍♀️

**Comment by mpeyper** [comment]
This no longer appears to be an issue in `lerna` version `3.7.1` (I'm not sure which release after `3.4.3` fixed it).

**Comment by deporter** [comment]
It’s no longer an issue for me either with the latest lerna and updated npm.

**Comment by respectTheCode** [comment]
I am also running into this issue. I can publish with NPM but not Lerna

```lerna ERR! EACCESS You do not have write permission required to publish "@org/package"```

**Comment by evocateur** [comment]
Lerna hasn't used `npm` in a suprocess to publish for quite some time. Please open a new issue.
